EVEREST-213 Added docs (#80)
percona/percona-everest-backend: CONTRIBUTING.md

# Contributing to Percona Everest backend

Percona Everest backend uses two types of methods:

- "own" methods, such as registering a Kubernetes cluster in Everest and listing the clusters.
-  proxy methods for the Kubernetes API, including all resource-related methods like database-cluster, database-cluster-restore, and database-engine.

The API server basic code is generated using [oapi-codegen](https://github.com/deepmap/oapi-codegen) from the docs/spec/openapi.yml file.
The proxy methods align with Everest operator methods but don't support all original parameters, because these are not required.

> You can find the definition of the custom resources in the [Everest operator repo](https://github.com/percona/dbaas-operator/tree/main/config/crd/bases).

Reviewer: outdated link to the dbaas-operator repo

### Run percona-everest-backend locally
0. Prerequisites:
    - Golang 1.20.x
    - Make 3.x
    - Docker 20.x
    - Git 2.x
1. Check out the repo:
`git clone https://github.com/percona/percona-everest-backend`
2. Navigate to the repo folder:
`cd percona-everest-backend`
3. Check out a particular branch if needed:
`git checkout <branch_name>`
4. Install the project dependencies: 
`make init`
5. Run the dev environment: 
`make local-env-up`
6. Run the build: `make run`

### Add a new proxy method
1. Copy the corresponding k8s spec to the [openapi.yml](./docs/spec/openapi.yml). For information on observing your cluster API, see [Kubernetes: How to View Swagger UI blog post](https://jonnylangefeld.com/blog/kubernetes-how-to-view-swagger-ui), which details the operator-defined methods (if the operator is installed).

> 

Reviewer: To avoid confusions, let's refer to the Everest Operator, not just an operator.

2. Make necessary spec modifications. When designing new methods:

-  follow the [Restful API guidelines](https://opensource.zalando.com/restful-api-guidelines/). - - use kebab-case instead of operator API. 

> - determine parameters to expose via proxy.```

Reviewer: the  \```  is a leftover

> 3. If needed, copy the custom resources schema from the [Everest operator config](https://github.com/percona/dbaas-operator/tree/main/config/crd/bases) to the **Components** section of the [openapi.yml](./docs/spec/openapi.yml) file.

Reviewer: outdated link to the dbaas-operator repo

# Contributing to Percona Everest backend

> Percona Everest backend uses two types of methods:

Reviewer 2: ```suggestion
Percona Everest backend uses two types of methods:

```

- "own" methods, such as registering a Kubernetes cluster in Everest and listing the clusters.
-  proxy methods for the Kubernetes API, including all resource-related methods like database-cluster, database-cluster-restore, and database-engine.

The API server basic code is generated using [oapi-codegen](https://github.com/deepmap/oapi-codegen) from the docs/spec/openapi.yml file.

> 

Reviewer 2: ```suggestion
```

### Run percona-everest-backend locally
0. Prerequisites:
    - Golang 1.20.x
    - Make 3.x
    - Docker 20.x
    - Git 2.x
1. Check out the repo:
`git clone https://github.com/percona/percona-everest-backend`
2. Navigate to the repo folder:
`cd percona-everest-backend`
3. Check out a particular branch if needed:
`git checkout <branch_name>`
4. Install the project dependencies: 
`make init`
5. Run the dev environment: 
`make local-env-up`
6. Run the build: `make run`

### Add a new proxy method
1. Copy the corresponding k8s spec to the [openapi.yml](./docs/spec/openapi.yml). For information on observing your cluster API, see [Kubernetes: How to View Swagger UI blog post](https://jonnylangefeld.com/blog/kubernetes-how-to-view-swagger-ui), which details the operator-defined methods (if the operator is installed).

2. Make necessary spec modifications. When designing new methods:

-  follow the [Restful API guidelines](https://opensource.zalando.com/restful-api-guidelines/). - - use kebab-case instead of operator API. 
- determine parameters to expose via proxy.```

> 2. If needed, copy the custom resources schema from the [Everest operator config](https://github.com/percona/dbaas-operator/tree/main/config/crd/bases) to the **Components** section of the [openapi.yml](./docs/spec/openapi.yml) file.

Reviewer 2: ```suggestion
3. If needed, copy the custom resources schema from the [Everest operator config](https://github.com/percona/dbaas-operator/tree/main/config/crd/bases) to the **Components** section of the [openapi.yml](./docs/spec/openapi.yml) file.
```

> 3. Run the following command to generate the code:

Reviewer 2: ```suggestion
4. Run the following command to generate the code:
```

```
 $ make init
 $ make gen
```

> 4. Implement the missing `ServerInterface` methods.

Reviewer 2: ```suggestion
5. Implement the missing `ServerInterface` methods.
```

> 5. Run `make format` to format the code and group the imports.

Reviewer 2: ```suggestion
6. Run `make format` to format the code and group the imports.
```

> 6. Run `make check` to verify that your code works and meets all style requirements.

Reviewer 2: ```suggestion
7. Run `make check` to verify that your code works and meets all style requirements.
```

### Running integration tests 

To run integration tests, see [Percona Everest API integration tests](api-tests/README.md).

### Working with local Kubernetes instances like Minikube or Kind 

When working with local Kubernetes clusters, Everest backend cannot connect to them because they often use `127.0.0.1` or `localhost` addresses. However, it is possible to connect to the host machine using `host.docker.internal` hostname since Everest backend runs inside a Docker container.

> 

Reviewer 2: ```suggestion
```

### Running integration tests 

Please follow the guideline [here](api-tests/README.md)

### Working with local kubernetes instances like minikube or kind 

The main issue you can face while working with local kubernetes clusters is that everest backend can't connect to those clusters because usually they use `127.0.0.1` or `localhost` addresses. Everest backend runs inside docker container and there's a way to connect to the host machine using `host.docker.internal` hostname

On your local machine you need to add this to `/etc/hosts` file

```
127.0.0.1          host.docker.internal
```

#### Running minikube clusters
To spin-up minikube cluster depending on your operating system you need to provide `--apiserver-names host.docker.internal`

We have `make local-env-up` command available in [everest-operator](https://github.com/percona/everest-operator/blob/main/Makefile#L301) and you can use it. It works fine on MacOS.

Once you kubeconfig will be available for your minikube cluster you need to rewrite server address in kubeconfig from `127.0.0.1` to `host.docker.internal` keeping port.

After that you can use everest by running provisioning from CLI

### Troubleshooting

> Some commands might help you understand what's going wrong

Reviewer 2: ```suggestion
Here are some commands that can help you fix potential issues:
```
